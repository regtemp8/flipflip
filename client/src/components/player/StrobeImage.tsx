import React, { type PropsWithChildren } from 'react'
import { animated, useSpring } from 'react-spring'

import { getDuration, getEaseFunction } from '../../data/utils'
import { useAppSelector } from '../../store/hooks'
import {
  selectSceneStrobeTF,
  selectSceneStrobeDuration,
  selectSceneStrobeDurationMin,
  selectSceneStrobeDurationMax,
  selectSceneStrobeSinRate,
  selectSceneStrobeBPMMulti,
  selectSceneTransEase,
  selectSceneTransExp,
  selectSceneTransAmp,
  selectSceneTransPer,
  selectSceneTransOv
} from '../../store/scene/selectors'
import { selectAudioBPM } from '../../store/audio/selectors'
import { selectUndefined } from '../../store/app/selectors'

export interface StrobeImageProps {
  sceneID: number
  timeToNextFrame: number
  currentAudio?: number
}

export default function StrobeImage(
  props: PropsWithChildren<StrobeImageProps>
) {
  const strobeTF = useAppSelector(selectSceneStrobeTF(props.sceneID))
  const strobeTime = useAppSelector(selectSceneStrobeDuration(props.sceneID))
  const strobeTimeMin = useAppSelector(
    selectSceneStrobeDurationMin(props.sceneID)
  )
  const strobeTimeMax = useAppSelector(
    selectSceneStrobeDurationMax(props.sceneID)
  )
  const strobeSinRate = useAppSelector(selectSceneStrobeSinRate(props.sceneID))
  const strobeBPMMulti = useAppSelector(
    selectSceneStrobeBPMMulti(props.sceneID)
  )
  const transEase = useAppSelector(selectSceneTransEase(props.sceneID))
  const transExp = useAppSelector(selectSceneTransExp(props.sceneID))
  const transAmp = useAppSelector(selectSceneTransAmp(props.sceneID))
  const transPer = useAppSelector(selectSceneTransPer(props.sceneID))
  const transOv = useAppSelector(selectSceneTransOv(props.sceneID))
  const bpmSelector =
    props.currentAudio != null
      ? selectAudioBPM(props.currentAudio)
      : selectUndefined
  const bpm = useAppSelector(bpmSelector)

  const duration = getDuration(
    {
      timingFunction: strobeTF,
      time: strobeTime,
      timeMax: strobeTimeMax,
      timeMin: strobeTimeMin,
      sinRate: strobeSinRate,
      bpmMulti: strobeBPMMulti
    },
    props.timeToNextFrame,
    bpm
  )

  const imageProps = useSpring({
    reset: true,
    from: {
      opacity: 1
    },
    to: {
      opacity: 0
    },
    config: {
      duration,
      easing: getEaseFunction(transEase, transExp, transAmp, transPer, transOv)
    }
  })

  return (
    <animated.div
      style={{
        position: 'absolute',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        overflow: 'hidden',
        zIndex: 2,
        ...imageProps
      }}
    >
      {props.children}
    </animated.div>
  )
}

;(StrobeImage as any).displayName = 'StrobeImage'
