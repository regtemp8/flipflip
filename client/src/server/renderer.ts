import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
