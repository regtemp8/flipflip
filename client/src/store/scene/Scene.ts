import type AudioPlaylist from './AudioPlaylist'
import type ScriptPlaylist from './ScriptPlaylist'
import type WeightGroup from './WeightGroup'
import {
  copy,
  EA,
  TF,
  BT,
  IF,
  IT,
  GO,
  OF,
  OT,
  VO,
  WF,
  SOF,
  HTF,
  VTF,
  STF,
  SL,
  SC
} from 'flipflip-common'

export default interface Scene {
  [key: string]:
    | number
    | number[]
    | number[][]
    | string
    | string[]
    | boolean
    | AudioPlaylist[]
    | ScriptPlaylist[]
    | WeightGroup[]
    | undefined

  id: number
  name: string
  sources: number[] // Array of LibrarySource IDs
  useWeights: boolean

  timingFunction: string
  timingConstant: number
  timingMin: number
  timingMax: number
  timingSinRate: number
  timingBPMMulti: number
  backForth: boolean
  backForthTF: string
  backForthConstant: number
  backForthMin: number
  backForthMax: number
  backForthSinRate: number
  backForthBPMMulti: number
  imageType: string
  backgroundType: string
  backgroundColor: string
  backgroundColorSet: string[]
  backgroundBlur: number

  imageTypeFilter: string
  fullSource: boolean
  imageOrientation: string
  gifOption: string
  gifTimingConstant: number
  gifTimingMin: number
  gifTimingMax: number
  videoOrientation: string
  videoOption: string
  videoTimingConstant: number
  videoTimingMin: number
  videoTimingMax: number
  videoSpeed: number
  videoRandomSpeed: boolean
  videoSpeedMin: number
  videoSpeedMax: number
  videoSkip: number
  randomVideoStart: boolean
  continueVideo: boolean
  playVideoClips: boolean
  skipVideoStart: number
  skipVideoEnd: number
  videoVolume: number
  weightFunction: string
  sourceOrderFunction: string
  forceAllSource: boolean
  orderFunction: string
  forceAll: boolean

  zoom: boolean
  zoomRandom: boolean
  zoomStart: number
  zoomStartMin: number
  zoomStartMax: number
  zoomEnd: number
  zoomEndMin: number
  zoomEndMax: number
  horizTransType: string
  horizTransLevel: number
  horizTransLevelMin: number
  horizTransLevelMax: number
  horizTransRandom: boolean
  vertTransType: string
  vertTransLevel: number
  vertTransLevelMin: number
  vertTransLevelMax: number
  vertTransRandom: boolean
  transTF: string
  transDuration: number
  transDurationMin: number
  transDurationMax: number
  transSinRate: number
  transBPMMulti: number
  transEase: string
  transExp: number
  transAmp: number
  transPer: number
  transOv: number

  crossFade: boolean
  crossFadeAudio: boolean
  fadeTF: string
  fadeDuration: number
  fadeDurationMin: number
  fadeDurationMax: number
  fadeSinRate: number
  fadeBPMMulti: number
  fadeEase: string
  fadeExp: number
  fadeAmp: number
  fadePer: number
  fadeOv: number

  slide: boolean
  slideTF: string
  slideType: string
  slideDistance: number
  slideDuration: number
  slideDurationMin: number
  slideDurationMax: number
  slideSinRate: number
  slideBPMMulti: number
  slideEase: string
  slideExp: number
  slideAmp: number
  slidePer: number
  slideOv: number

  strobe: boolean
  strobePulse: boolean
  strobeLayer: string
  strobeOpacity: number
  strobeTF: string
  strobeTime: number
  strobeTimeMin: number
  strobeTimeMax: number
  strobeSinRate: number
  strobeBPMMulti: number
  strobeDelayTF: string
  strobeDelay: number
  strobeDelayMin: number
  strobeDelayMax: number
  strobeDelaySinRate: number
  strobeDelayBPMMulti: number
  strobeColorType: string
  strobeColor: string
  strobeColorSet: string[]
  strobeEase: string
  strobeExp: number
  strobeAmp: number
  strobePer: number
  strobeOv: number

  fadeInOut: boolean
  fadeIOPulse: boolean
  fadeIOTF: string
  fadeIODuration: number
  fadeIODurationMin: number
  fadeIODurationMax: number
  fadeIOSinRate: number
  fadeIOBPMMulti: number
  fadeIODelayTF: string
  fadeIODelay: number
  fadeIODelayMin: number
  fadeIODelayMax: number
  fadeIODelaySinRate: number
  fadeIODelayBPMMulti: number
  fadeIOStartEase: string
  fadeIOStartExp: number
  fadeIOStartAmp: number
  fadeIOStartPer: number
  fadeIOStartOv: number
  fadeIOEndEase: string
  fadeIOEndExp: number
  fadeIOEndAmp: number
  fadeIOEndPer: number
  fadeIOEndOv: number

  panning: boolean
  panTF: string
  panDuration: number
  panDurationMin: number
  panDurationMax: number
  panSinRate: number
  panBPMMulti: number
  panHorizTransType: string
  panHorizTransImg: boolean
  panHorizTransLevel: number
  panHorizTransLevelMax: number
  panHorizTransLevelMin: number
  panHorizTransRandom: boolean
  panVertTransType: string
  panVertTransImg: boolean
  panVertTransLevel: number
  panVertTransLevelMax: number
  panVertTransLevelMin: number
  panVertTransRandom: boolean
  panStartEase: string
  panStartExp: number
  panStartAmp: number
  panStartPer: number
  panStartOv: number
  panEndEase: string
  panEndExp: number
  panEndAmp: number
  panEndPer: number
  panEndOv: number

  overrideIgnore: boolean
  gridScene: boolean
  scriptScene: boolean
  downloadScene: boolean
  generatorMax: number
  overlayEnabled: boolean
  overlays: number[] // Array of Overlay IDs
  nextSceneID: number
  nextSceneTime: number
  nextSceneAllImages: boolean
  persistAudio: boolean
  persistText: boolean
  nextSceneRandomID: number
  nextSceneRandoms: number[]
  libraryID: number
  audioScene: boolean
  audioEnabled: boolean
  audioPlaylists: number[]
  audioStartIndex: number
  textEnabled: boolean
  scriptPlaylists: number[]
  scriptStartIndex: number
  regenerate: boolean
  generatorWeights?: WeightGroup[]
  openTab: number
}

export const initialScene: Scene = {
  id: 0,
  name: 'Unnamed scene',
  sources: [],
  useWeights: false,
  timingFunction: TF.constant,
  timingConstant: 1000,
  timingMin: 200,
  timingMax: 1200,
  timingSinRate: 100,
  timingBPMMulti: 10,
  backForth: false,
  backForthTF: TF.constant,
  backForthConstant: 1000,
  backForthMin: 200,
  backForthMax: 1200,
  backForthSinRate: 100,
  backForthBPMMulti: 10,
  imageType: IT.fitBestNoClip,
  backgroundType: BT.blur,
  backgroundColor: '#000000',
  backgroundColorSet: [],
  backgroundBlur: 8,
  imageTypeFilter: IF.any,
  fullSource: false,
  imageOrientation: OT.original,
  gifOption: GO.none,
  gifTimingConstant: 3000,
  gifTimingMin: 1000,
  gifTimingMax: 3000,
  videoOrientation: OT.original,
  videoOption: VO.none,
  videoTimingConstant: 3000,
  videoTimingMin: 1000,
  videoTimingMax: 3000,
  videoSpeed: 10,
  videoRandomSpeed: false,
  videoSpeedMin: 5,
  videoSpeedMax: 20,
  videoSkip: 10,
  randomVideoStart: false,
  continueVideo: false,
  playVideoClips: true,
  skipVideoStart: 0,
  skipVideoEnd: 0,
  videoVolume: 0,
  weightFunction: WF.sources,
  sourceOrderFunction: SOF.random,
  forceAllSource: false,
  orderFunction: OF.random,
  forceAll: false,
  zoom: false,
  zoomRandom: false,
  zoomStart: 1,
  zoomStartMin: 0.5,
  zoomStartMax: 1,
  zoomEnd: 2,
  zoomEndMin: 1.5,
  zoomEndMax: 2,
  horizTransType: HTF.none,
  horizTransLevel: 10,
  horizTransLevelMin: 5,
  horizTransLevelMax: 10,
  horizTransRandom: false,
  vertTransType: VTF.none,
  vertTransLevel: 10,
  vertTransLevelMin: 5,
  vertTransLevelMax: 10,
  vertTransRandom: false,
  transTF: TF.constant,
  transDuration: 5000,
  transDurationMin: 1000,
  transDurationMax: 7000,
  transSinRate: 100,
  transBPMMulti: 10,
  transEase: EA.linear,
  transExp: 6,
  transAmp: 20,
  transPer: 6,
  transOv: 3,
  crossFade: false,
  crossFadeAudio: false,
  fadeTF: TF.constant,
  fadeDuration: 500,
  fadeDurationMin: 100,
  fadeDurationMax: 700,
  fadeSinRate: 100,
  fadeBPMMulti: 10,
  fadeEase: EA.linear,
  fadeExp: 6,
  fadeAmp: 20,
  fadePer: 6,
  fadeOv: 3,
  slide: false,
  slideTF: TF.constant,
  slideType: STF.left,
  slideDistance: 100,
  slideDuration: 500,
  slideDurationMin: 100,
  slideDurationMax: 700,
  slideSinRate: 100,
  slideBPMMulti: 10,
  slideEase: EA.linear,
  slideExp: 6,
  slideAmp: 20,
  slidePer: 6,
  slideOv: 3,
  strobe: false,
  strobePulse: false,
  strobeLayer: SL.top,
  strobeOpacity: 1,
  strobeTF: TF.constant,
  strobeTime: 200,
  strobeTimeMin: 100,
  strobeTimeMax: 300,
  strobeSinRate: 100,
  strobeBPMMulti: 10,
  strobeDelayTF: TF.constant,
  strobeDelay: 200,
  strobeDelayMin: 100,
  strobeDelayMax: 300,
  strobeDelaySinRate: 100,
  strobeDelayBPMMulti: 10,
  strobeColorType: SC.color,
  strobeColor: '#FFFFFF',
  strobeColorSet: [],
  strobeEase: EA.linear,
  strobeExp: 6,
  strobeAmp: 20,
  strobePer: 6,
  strobeOv: 3,
  fadeInOut: false,
  fadeIOPulse: false,
  fadeIOTF: TF.constant,
  fadeIODuration: 2000,
  fadeIODurationMin: 2000,
  fadeIODurationMax: 5000,
  fadeIOSinRate: 100,
  fadeIOBPMMulti: 10,
  fadeIODelayTF: TF.constant,
  fadeIODelay: 2000,
  fadeIODelayMin: 2000,
  fadeIODelayMax: 5000,
  fadeIODelaySinRate: 100,
  fadeIODelayBPMMulti: 10,
  fadeIOStartEase: EA.linear,
  fadeIOStartExp: 6,
  fadeIOStartAmp: 20,
  fadeIOStartPer: 6,
  fadeIOStartOv: 3,
  fadeIOEndEase: EA.linear,
  fadeIOEndExp: 6,
  fadeIOEndAmp: 20,
  fadeIOEndPer: 6,
  fadeIOEndOv: 3,
  panning: false,
  panTF: TF.constant,
  panDuration: 2000,
  panDurationMin: 2000,
  panDurationMax: 5000,
  panSinRate: 100,
  panBPMMulti: 10,
  panHorizTransType: HTF.none,
  panHorizTransImg: false,
  panHorizTransLevel: 10,
  panHorizTransLevelMax: 10,
  panHorizTransLevelMin: 5,
  panHorizTransRandom: false,
  panVertTransType: VTF.none,
  panVertTransImg: false,
  panVertTransLevel: 10,
  panVertTransLevelMax: 10,
  panVertTransLevelMin: 5,
  panVertTransRandom: false,
  panStartEase: EA.linear,
  panStartExp: 6,
  panStartAmp: 20,
  panStartPer: 6,
  panStartOv: 3,
  panEndEase: EA.linear,
  panEndExp: 6,
  panEndAmp: 20,
  panEndPer: 6,
  panEndOv: 3,
  overrideIgnore: false,
  gridScene: false,
  scriptScene: false,
  downloadScene: false,
  generatorMax: 100,
  overlayEnabled: false,
  overlays: [],
  nextSceneID: 0,
  nextSceneTime: 900000,
  nextSceneAllImages: false,
  persistAudio: false,
  persistText: false,
  nextSceneRandomID: 0,
  nextSceneRandoms: [],
  libraryID: -1,
  audioScene: false,
  audioEnabled: false,
  audioPlaylists: [],
  audioStartIndex: 0,
  textEnabled: false,
  scriptPlaylists: [],
  scriptStartIndex: 0,
  regenerate: true,
  openTab: 3
}

export function newScene(init?: Partial<Scene>): Scene {
  return Object.assign(copy<Scene>(initialScene), init)
}
