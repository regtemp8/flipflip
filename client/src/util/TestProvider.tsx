import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
