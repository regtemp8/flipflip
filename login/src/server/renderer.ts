import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
