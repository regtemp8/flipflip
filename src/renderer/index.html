<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
