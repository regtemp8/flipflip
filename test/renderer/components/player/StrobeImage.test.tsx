import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import StrobeImage from "../../../../src/renderer/components/player/StrobeImage";
import TestProvider from "../../../util/TestProvider";
import Scene from "../../../../src/renderer/data/Scene";

describe("StrobeImage", () => {
  it("should match snapshot", () => {
    const scene = new Scene();
    const component = renderer.create(
      <TestProvider>
        <StrobeImage scene={scene} timeToNextFrame={0} currentAudio={null}>
          <p>Test</p>
        </StrobeImage>
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
