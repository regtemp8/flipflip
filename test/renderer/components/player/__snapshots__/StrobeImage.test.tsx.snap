// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrobeImage should match snapshot 1`] = `
<div
  style={
    {
      "bottom": 0,
      "left": 0,
      "opacity": 1,
      "overflow": "hidden",
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 2,
    }
  }
>
  <p>
    Test
  </p>
</div>
`;
